content/md/posts/clojure-which-execution-option-to-use.md

```diff
@@ -0,0 +1,217 @@
+
+
+## clojure.exec -X
+
+`-X` flag provides the flexibility to call any fully qualified function, so Clojure code is no longer tied to `-main`
+
+
+
+Arguments are passed as EDN key value pairs and can therefore be included in any order
+
+> The namesapce of the function to be called is automatically loaded using `requiring-resolve!` (TODO: explanation)
+
+Thus you can now define programs as functions that take a map and execute them via clj:
+
+Call the `status` function from the namespace `practicalli.service` in the current project
+
+```bash
+clojure -X practicalli.service/status
+```
+
+Pass arguments to a `start` function in the `practicalli.service` namespace
+
+```bash
+clojure -X practicalli.service/status :port 8080 :join? false
+```
+
+As the arguments are key/value pairs, it does not matter in which order the pairs are used in the command line.
+
+### Built in clojure.exec functions
+Clojure CLI tools has some built in tools under the special `:deps` alias (not to be confused with the `:deps` configuration in a `deps.edn` file)
+
+* `-X:deps mvn-install` - install a maven jar to the local repository cache
+* `-X:deps git-resolve-tags` - Resolve git coord tags to shas and update deps.edn
+* `-X:deps find-versions` - Find available versions of a library
+* `-X:deps prep` - [prepare source code libraries](https://clojure.org/reference/deps_and_cli#prep) in the dependency tree
+
+> See `clojure --help` for other descriptions and examples (or just read this article further)
+
+
+## Tools -T
+
+Install, run and remove a tool that can be used independently from a Clojure project configuration.
+
+The `-T` flag is based on the `clojure.exec` approach, except it only uses the `:extra-deps` and `:extra-path` from the alias, ignoring the `:deps` and `:path` values from the project `deps.edn` file.
+
+`-Ttools` is a built in tool for `install` and `remove` of other tools.
+
+The  `:as` directive specifying a name for the too
+l
+
+In this example, the antq tool is installed using the name `libs-outdated`
+
+```bash
+clojure -Ttools install com.github.liquidz/antq '{:git/tag "1.3.1"}' :as libs-outdated
+```
+
+To update a tool, simply run the install again with the latest version of the library or git reference.
+
+Installing a tool adds an EDN configuration file using the name of the tool in `$HOME/.clojure/tools/` (or $XDG_HOME/.clojure/tools/) directory.
+
+Once a tool is installed, run by using the name of the tool.
+
+```bash
+clojure -Tlibs-outdated
+```
+
+Options to the tool are passed as key/value pairs (as the tool is called by clojure.exec)
+
+```bash
+clojure -Tlibs-outdated :upgrade true
+```
+
+`-Ttools remove` will remove the configuration of the tool of the given name
+
+```bash
+clojure -Ttools remove :tool libs-outdated
```

Comment:
Don't forget to update this line if you rename the installed version per above review comments.